http/apache/nanny.py

```diff
@@ -0,0 +1,318 @@
+#! /usr/bin/env python3
+"""Load balancer around apache."""
+import argparse
+import asyncio
+import concurrent.futures
+import contextlib
+import dataclasses
+import functools
+import http.client
+import itertools
+import json
+import logging
+import os
+import pathlib
+import socket
+import shutil
+import subprocess
+from typing import Any, Generator, List, TypeVar
+
+import psutil
+
+T = TypeVar('T')
+
+logger = logging.getLogger()
+
+APACHE_RUN_DIR = '/var/run/apache2'
+APACHE_LOG_DIR = '/var/log/apache2'
+APACHE_LOCK_DIR = '/var/lock/apache2'
+
+
+@dataclasses.dataclass
+class ApacheInstance:
+    port: int
+    log_dir: pathlib.Path
+    pid_file_path: pathlib.Path
+
+
+class NonClosingResponse(http.client.HTTPResponse):
+    def _close_conn(self):
+        # Normally this would close the file obj, but since we're using
+        # pipelining, we need to read multiple responses from the same
+        # connection.
+        pass
+
+
+class MockSocket:
+    """Mocket.
+
+    HTTPResponse only uses the ``makefile`` method of sockets.
+    """
+    def __init__(self, fp):
+        self.fp = fp
+    def makefile(self, *args, **kwargs):
+        return self.fp
+
+
+async def read_with_prepended_length(reader: asyncio.StreamReader) -> bytes:
+    """Raises IncompleteReadError if not all data is found."""
+    expected_size_str = await reader.readuntil(b'\n')
+    expected_size = int(expected_size_str)
+    return await reader.readexactly(expected_size)
+
+
+@contextlib.contextmanager
+def reenqueueing(resource: T, queue: "asyncio.Queue[T]") -> Generator[T, Any, None]:
+    try:
+        yield resource
+    finally:
+        queue.put_nowait(resource)
+
+
+async def handle_request_stream(
+    reader: asyncio.StreamReader,
+    writer: asyncio.StreamWriter,
+    *,
+    apache_instance_queue: "asyncio.Queue[ApacheInstance]",
+) -> None:
+    logger.debug(f'Request {id(reader):x}: Received')
+    loop = asyncio.get_running_loop()
+    with reenqueueing(await apache_instance_queue.get(), apache_instance_queue) as apache_instance:
+        logger.debug(f'Request {id(reader):x}: Claimed server instance on port {apache_instance.port}')
+        ensure_instance_running(apache_instance)
+        results_log_path = apache_instance.log_dir / 'results.log'
+        error_log_path = apache_instance.log_dir / 'error.log'
+        access_log_path = apache_instance.log_dir / 'access.log'
+        # empty logs before sending request stream
+        for log_path in [results_log_path, error_log_path, access_log_path]:
+            open(log_path, 'w').close()
```

Comment:
Context manager would be preferable for tightness

Reply:
E.g.:

```python
with open(log_path, 'w'):
    pass
```

Reply:
@caleblf

Reply:
I wish there were a more explicit function in the stdlib to blank a file. Neither of these options seems ideal. But yeah, sure, might as well use `with`.